tests/lib/fakestore/store/store.go

```diff
@@ -739,49 +740,109 @@ func (s *Store) retrieveAssertion(bs asserts.Backstore, assertType *asserts.Asse
 	return a, err
 }
 
-func (s *Store) assertionsEndpoint(w http.ResponseWriter, req *http.Request) {
-	assertPath := strings.TrimPrefix(req.URL.Path, "/v2/assertions/")
+func (s *Store) retrieveSequenceFormingAssertion(bs asserts.Backstore, assertType *asserts.AssertionType, sequenceKey []string, sequence int) (asserts.Assertion, error) {
+	if sequence > 0 {
+		hdrs, err := asserts.HeadersFromSequenceKey(assertType, sequenceKey)
+		if err != nil {
+			return nil, fmt.Errorf("%q assertion reference sequence key %v is invalid: %v", assertType.Name, sequenceKey, err)
+		}
+		return nil, &asserts.NotFoundError{
+			Type:    assertType,
+			Headers: hdrs,
+		}
+	}
 
-	bs, err := s.collectAssertions()
-	if err != nil {
-		http.Error(w, fmt.Sprintf("internal error collecting assertions: %v", err), 500)
-		return
+	a, err := bs.SequenceMemberAfter(assertType, sequenceKey, -1, assertType.MaxSupportedFormat())
+	if errors.Is(err, &asserts.NotFoundError{}) && s.assertFallback {
+		return s.fallback.SeqFormingAssertion(assertType, sequenceKey, sequence, nil)
 	}
+	return a, err
+}
 
-	comps := strings.Split(assertPath, "/")
+func (s *Store) sequenceFromQueryValues(values url.Values) (int, error) {
+	if val, ok := values["sequence"]; ok {
+		// special case value of 'latest', in that case
+		// we return -1 to indicate we want the newest
+		if val[0] != "latest" {
+			seq, err := strconv.Atoi(val[0])
+			if err != nil {
+				return -1, fmt.Errorf("cannot parse sequence %s: %v", val[0], err)
+			}
+			return seq, nil
+		}
+	}
+	return -1, nil
+}
 
+func (s *Store) assertTypeAndKey(urlPath string) (*asserts.AssertionType, []string, error) {
+	// trim the assertions prefix, and handle any query parameters
+	assertPath := strings.TrimPrefix(urlPath, "/v2/assertions/")
+	comps := strings.Split(assertPath, "/")
 	if len(comps) == 0 {
-		http.Error(w, "missing assertion type", 400)
-		return
+		return nil, nil, fmt.Errorf("missing assertion type")
 	}
 
 	typ := asserts.Type(comps[0])
 	if typ == nil {
-		http.Error(w, fmt.Sprintf("unknown assertion type: %s", comps[0]), 400)
+		return nil, nil, fmt.Errorf("unknown assertion type: %s", comps[0])
+	}
+	return typ, comps[1:], nil
+}
+
+func (s *Store) retrieveAssertionWrapper(bs asserts.Backstore, assertType *asserts.AssertionType, keyParts []string, values url.Values) (asserts.Assertion, error) {
+
+	pk := keyParts
+	if assertType.SequenceForming() {
+		seq, err := s.sequenceFromQueryValues(values)
+		if err != nil {
+			return nil, err
+		}
+
+		// If no sequence key is present, then we use a special lookup to
```

Comment:
Looks like it should perhaps be 'No sequence key or sequence key "latest"' will result in the -1 special lookup?

Reply:
This makes a lot of sense, thank you!